ExodusMovement/asana-actions
# feat: set Asana to milestone based on PR milestone

## Summary



This PR sets the _milestone_ in the asana tasks' linked in the PR.

Once a PR has at least one Asana task linked, and a milestone is set to the PR, the automation will run and will try to set the same milestone in the task that in the PR.

e.g. (covered by tests):
1. A PR with no Asana task is set with a new milestone -> the automation does nothing.
2. A PR with one Asana task linked is set with a new milestone -> the automation will search for the milestone in Asana and will set the milestone in the task.
3. (same as above) but the milestone is not found in Asana -> does nothing.
4. Multiple tasks are found in the PR and a milestone is set -> for all of them try to set the milestone
5. A PR has one or multiple tasks linked, a milestone is set but the tasks already have a milestone set -> the automation overrides the milestone.

Note:
The automation will assume that:
- _Milestone_ field in Asana is interpreted by the automation by `Target Release Version` field.
  - This can be overwritten by setting `asana_milestone_field_name` in the GitHub action.
- Milestones in GH are highlighted by numbers (e.g. v1.0.0 === V1.0.0 === foo1.0.0)
  - This can be overwritten by setting `gh_milestone_regex` in the GH action.
- Milestones in Asana are highlighted by numbers (e.g. v1.0.0 === V1.0.0 === foo1.0.0)
  - This can be overwritten by setting `asana_milestone_regex` in the GH action.



Closes: [this Asana task](https://app.asana.com/0/1199981457200002/1202777284092475).

## Test Plan

- Checkout this branch locally and create a new branch
- Add a dummy change in `asana-actions-workflow` (a console.log for example)
- Commit and push changes
- Create an asana task for testing purposes 
- Create a PR from your branch to `federico/link-asana-milestone`
  - Copy-paste the asana task link after the `Closes:` prefix.
- After your PR is created, wait for a few seconds and the automation should have:
  - Pushed a comment in the task with the PR
  - Updated PR description with the comment
- Assign a milestone to the PR
  - Ensure the milestone matches with asana task (e.g. v22.08.11)
- Wait for a few seconds and the task should update with the milestone
- Finally, merge the PR and the task should mark as completed.

[DEMO](https://www.loom.com/share/167f85d27a3b4345b73c07bbbb974b89)

## Further Comments

Ok GH actions are a bit tricky. `Milestoned` and `Demilestoned` are triggered at the same time when you switch from Milestone X to Y, so they get overwritten each other and usually the tasks end up with `null` (empty milestone). I added a delay of 15 secs for milestoning the task. This ensures that milestoning tasks await for some secs until demilestone happened (in case it's a re-milestoning)
